### PR TITLE
Assignment 4 - Laura Martín

### PR DESCRIPTION
**Pull request by LLLP11**
task 6 and 7

### CODE REVIEW

File: LLLP11 - 22C005/task06.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+# Visualize the results
+g.add((ns.Researcher, RDFS.subClassOf, ns.Person))
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.3: Create a new individual of Researcher named "Jane Smith"**"""
+
+# TO DO
+# Visualize the results
+g.add((ns.JaneSmith, RDFS.Literal, ns.Researcher))
```
Comment (dgarijo) [inline LLLP11 - 22C005/task06.py:53]: This is incorrect. RDFS.Literal is not used to create individuals

File: LLLP11 - 22C005/task06.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+g.add((ns.JaneSmith, ns2.Given, Literal('Jane')))
+g.add((ns.JaneSmith, ns2.Family, Literal('Smith')))
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.5: Add UPM as the university where John Smith works**"""
+
+# TO DO
+# Visualize the results
+g.add((ns.UPM, RDFS.Literal, ns.University))
```
Comment (dgarijo) [inline LLLP11 - 22C005/task06.py:74]: wrong again

File: LLLP11 - 22C005/task06.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+g.add((ns.JohnSmith, ns2.works_at, ns.UPM))
+for s, p, o in g:
+  print(s,p,o)
+
+"""**Task 6.6: Add that Jown knows Jane using the FOAF vocabulary**"""
+
+# TO DO
+# Visualize the results
+FOAF = Namespace("http://xmlns.com/foaf/0.1/")
+g.add((ns.Jown, ns.knowsJaneuses, ns.FOAFvocabulary))
```
Comment (dgarijo) [inline LLLP11 - 22C005/task06.py:85]: I do not know what are you trying to do here.

File: LLLP11 - 22C005/task07.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+"""**TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**"""
+
+# TO DO
+# Visualize the results
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+vcard = Namespace("http://www.w3.orq/2001/vcard-rdf/3.0#")
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+  print(s)
+q1 = prepareQuery("SELECT ?x WHERE {?x rdfs:subClassOf ns:LivingThing}", initNs = {"rdfs": RDFS, "ns": ns})
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:33]: missing all subclasses of LivingThing

File: LLLP11 - 22C005/task07.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+"""
+
+# TO DO
+# Visualize the results
+for persona, p, o in g.triples((None,RDF.type, ns.Person)):
+  print(persona)
+for subclase, p ,o in g.triples((None, RDFS.subClassOf,ns.Person)):
+  for persona, p, o in g.triples((None, RDF.type,subclase)):
+    print(persona)
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:47]: Make the method recursive or you'll see a discrepancy in the responses

File: LLLP11 - 22C005/task07.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+# TO DO
+# Visualize the results
+for elem,p,o in g.triples((None, RDF.type, ns.Animal)):
+  for s,t,u in g.triples((elem, None,None)):
+    print(s,t,u)
+q1 = prepareQuery('''
+  SELECT ?x ?relacion
+  WHERE{
+    ?x rdf:type ns:Animal.
+    ?x ?relacion ?a
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:65]: What about the persons?

File: LLLP11 - 22C005/task07.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+# Visualize the results
+FOAF = Namespace("http://xmlns.com/foaf/0.1/")
+for persona,p,o in g.triples((None, RDF.type, ns.Person)):
+  for a,q,r in g.triples((None, FOAF.knows, persona)):
+    print(a)
+q1 = prepareQuery('''
+  SELECT DISTINCT ?x
+  WHERE{
+    ?x rdf:type ns:Person.
+    ?x FOAF:knows ns:RockySmith
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:83]: You are asked to list the name of the persons

File: LLLP11 - 22C005/task07.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+lista_def = []
+for n,s, o in g.triples((None, FOAF.knows, None)):
+  if s in lista:
+    lista_def += [s]
+  lista += [s]
+print(lista_def)
+q1 = prepareQuery('''
+  SELECT DISTINCT ?x
+  WHERE{
+    ?x FOAF:knows ?y
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:103]: You are not listing the entities that know at least two persons. You are listing the ones that know any person

**Review by dgarijo — CHANGES_REQUESTED** [review]
Please see pending changes to address

File: LLLP11 - 22C005/task06.py (hunk 1 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task06.py:53]: I've changed RDFS.Literal to RDF.type

g.add((ns.JaneSmith, **RDF.type**, ns.Researcher))
for s, p, o in g:
  print(s,p,o)

File: LLLP11 - 22C005/task06.py (hunk 2 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task06.py:74]: g.add((ns.UPM, **RDF.type,** ns.University))
g.add((ns2.works_at, RDF.type, RDFS.Datatype))
g.add((ns.JohnSmith, ns2.works_at, ns.UPM))
for s, p, o in g:
  print(s,p,o)

File: LLLP11 - 22C005/task06.py (hunk 3 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task06.py:85]: I've changed the code:
**foaf=Namespace("http://xmlns.com/foaf/0.1/")
g.add((ns.JohnSmith, RDF.type, foaf.Person))
g.add((ns.JaneSmith, RDF.type, foaf.Person))
g.add((ns.JohnSmith, foaf.knows, ns.JaneSmith))**
for s, p, o in g:
  print(s,p,o)

File: LLLP11 - 22C005/task07.py (hunk 4 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task07.py:33]: Added * to subClassOf (<- i can't put it in bold.)
ns:LivingThing -> http://somewhere#LivingThing (this http has <> but it makes they does not appear)

from rdflib.plugins.sparql import prepareQuery
ns = Namespace("http://somewhere#")
vcard = Namespace("http://www.w3.orq/2001/vcard-rdf/3.0#")
for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
  print(s)
q1 = prepareQuery("SELECT ?x WHERE {?x rdfs:subClassOf* <http://somewhere#LivingThing>.}", initNs = {"rdfs": RDFS, "ns": ns})
for r in g.query(q1):
  print(r)

File: LLLP11 - 22C005/task07.py (hunk 5 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task07.py:47]: q1 = prepareQuery('''
SELECT ?x ?y 
WHERE {
  {?x rdf:type/rdfs:subClassOf* ns:Person}
  UNION
  {?x rdf:type ns:Person}
  }''', initNs = {"rdf":RDF, "ns": ns})
for r in g.query(q1):
  print(r)

File: LLLP11 - 22C005/task07.py (hunk 6 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task07.py:65]: or elem,p,o in g.triples((None, RDF.type, ns.Animal)):
  for s,t,u in g.triples((elem, None,None)):
    print(s,t,u)
**for elem,p,o in g.triples((None, RDF.type, ns.Person)):
  for s,t,u in g.triples((elem, None,None)):
    print(s,t,u)**
q1 = prepareQuery('''
  SELECT ?x ?relacion
  WHERE {
    {?x rdf:type ns:Animal}
    {?x ?relacion ?a}
    **UNION
    {?x rdf:type ns:Person}
    {?x ?relacion ?a}**
    }
    ''', initNs = {"rdf":RDF, "ns": ns})
for r in g.query(q1):
  print(r)

File: LLLP11 - 22C005/task07.py (hunk 7 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task07.py:83]: Same problem with <> in <http://www.w3.org/2001/vcard-rdf/3.0/Given>


foaf = Namespace("http://xmlns.com/foaf/0.1/")
for persona,p,o in g.triples((None, RDF.type, ns.Person)):
  for a,q,r in g.triples((None, foaf.knows, persona)):
    print(a)
q1 = prepareQuery('''
 SELECT ?name
    WHERE {
        ?person foaf:knows ns:RockySmith.
        **?person <http://www.w3.org/2001/vcard-rdf/3.0/Given> ?name**
  }''', initNs = {"rdf": RDF, "ns": ns, "foaf": foaf}
  )
for r in g.query(q1):
  print(r)

File: LLLP11 - 22C005/task07.py (hunk 8 above)
Comment (LLLP11) [reply LLLP11 - 22C005/task07.py:103]: lista = []
lista_def = []
for n,s, o in g.triples((None, FOAF.knows, None)):
  if s in lista:
    lista_def += [s]
  lista += [s]
print(lista_def)
q1 = prepareQuery('''
  SELECT ?x ?person1 ?person2
  WHERE{
    ?x foaf:knows ?person1.
    ?x foaf:knows ?person2.
    FILTER(?person1 != ?person2)
  }''', initNs = {"rdf": RDF, "ns": ns, "FOAF":FOAF}
)
for r in g.query(q1):
  print(r)

**Comment by dgarijo** [comment]
Implement the changes in this PR, or open a new PR, reference this one and close this one.

File: LLLP11 - 22C005/task07.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+"""**TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**"""
+
+# TO DO
+# Visualize the results
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+vcard = Namespace("http://www.w3.orq/2001/vcard-rdf/3.0#")
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+    print(s)
+q1 = prepareQuery("SELECT ?x WHERE {?x rdfs:subClassOf* http://somewhere#LivingThing.}", initNs = {"rdfs": RDFS, "ns": ns})
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:33]: You have not run this query, it's malformed

File: LLLP11 - 22C005/task07.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+g.parse(github_storage+"/rdf/example6.rdf", format="xml")
+
+"""**TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**"""
+
+# TO DO
+# Visualize the results
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+vcard = Namespace("http://www.w3.orq/2001/vcard-rdf/3.0#")
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:31]: this does not retrieve all subclasses of subclasses of living thing

File: LLLP11 - 22C005/task07.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+"""
+
+# TO DO
+# Visualize the results
+q1 = prepareQuery('''
+    SELECT ?x ?y
+    WHERE {
+        {?x rdf:type/rdfs:subClassOf* ns:Person}
+        UNION
+        {?x rdf:type ns:Person}
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:48]:
```
UNION
        {?x rdf:type ns:Person}
``` 
is redundant

File: LLLP11 - 22C005/task07.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+# Visualize the results
+foaf = Namespace("http://xmlns.com/foaf/0.1/")
+for persona,p,o in g.triples((None, RDF.type, ns.Person)):
+    for a,q,r in g.triples((None, foaf.knows, persona)):
+        print(a)
+q1 = prepareQuery('''
+    SELECT ?name
+    WHERE {
+        ?person foaf:knows ns:RockySmith.
+        ?person http://www.w3.org/2001/vcard-rdf/3.0/Given ?name
```
Comment (dgarijo) [inline LLLP11 - 22C005/task07.py:93]: this query is malformed

**Review by dgarijo — CHANGES_REQUESTED** [review]
There are still issues

**Comment by anaigmo** [comment]
Entregado con algún fallo, no se admiten más PR

**Comment by anaigmo** [comment]
No sigue las instrucciones de entrega, así que no se hace merge

**Pull request closed without merging**